crates/derivation-pipeline/src/lib.rs

```diff
@@ -4,35 +4,71 @@
 //! into payload attributes for block building.
 
 #![cfg_attr(not(feature = "std"), no_std)]
+
+mod data_source;
+mod error;
+
 #[cfg(not(feature = "std"))]
 extern crate alloc as std;
 
 use std::vec::Vec;
 
+use crate::{data_source::CodecDataSource, error::DerivationPipelineError};
+use alloy_eips::eip4844::Blob;
 use alloy_primitives::B256;
 use alloy_rpc_types_engine::PayloadAttributes;
 use reth_scroll_chainspec::SCROLL_FEE_VAULT_ADDRESS;
+use rollup_node_primitives::BatchCommitData;
 use scroll_alloy_consensus::TxL1Message;
 use scroll_alloy_rpc_types_engine::ScrollPayloadAttributes;
-use scroll_codec::decoding::batch::Batch;
+use scroll_codec::Codec;
 
-/// An instance of the trait can be used to provide the next L1 message to be used in the derivation
-/// pipeline.
+/// An instance of the trait can provide L1 messages using a cursor approach. Set the cursor for the
+/// provider using the queue index or hash and then call [`L1MessageProvider::next_l1_message`] to
+/// iterate the queue.
 pub trait L1MessageProvider {
-    /// Returns the next L1 message.
+    /// Returns the L1 message at the current cursor and advances the cursor.
     fn next_l1_message(&self) -> TxL1Message;
+    /// Set the index cursor for the provider.
+    fn set_index_cursor(&mut self, index: u64);
+    /// Set the hash cursor for the provider.
+    fn set_hash_cursor(&mut self, hash: B256);
+}
+
+/// An instance of the trait can be used to provide L1 data.
+pub trait L1Provider: L1MessageProvider {
+    /// Returns corresponding blob data for the provided hash.
+    fn blob(&self, hash: B256) -> Option<Blob>;
 }
```

Comment:
Is there a reason that `blob` is at the root level and `L1MessageProvider` methods are not? How about:

```suggestion
pub trait L1Provider: L1MessageProvider + L1BlobProvider {}
impl L1Provider for T where T: L1MessageProvider + L1BlobProvider {}
```